network/src/gossip.rs

```diff
@@ -129,57 +207,586 @@ pub(crate) struct MessageValidationData {
-	live_session: RwLock<HashMap<Hash, MessageValidationData>>,
-	oracle: O,
+	report_handle: Box<Fn(&PeerId, i32) + Send + Sync>,
+	inner: RwLock<Inner<O>>,
+}
+
+impl<O: KnownOracle + ?Sized> MessageValidator<O> {
+	#[cfg(test)]
+	fn new_test(
+		oracle: O,
+		report_handle: Box<Fn(&PeerId, i32) + Send + Sync>,
+	) -> Self where O: Sized{
+		MessageValidator {
+			report_handle,
+			inner: RwLock::new(Inner {
+				peers: HashMap::new(),
+				our_view: Default::default(),
+				oracle,
+			})
+		}
+	}
+
+	fn report(&self, who: &PeerId, cost_benefit: i32) {
+		(self.report_handle)(who, cost_benefit)
+	}
 }
 
 impl<O: KnownOracle + ?Sized> network_gossip::Validator<Block> for MessageValidator<O> {
-	fn validate(&self, context: &mut ValidatorContext<Block>, _sender: &PeerId, mut data: &[u8])
+	fn new_peer(&self, _context: &mut ValidatorContext<Block>, who: &PeerId, _roles: Roles) {
+		let mut inner = self.inner.write();
+		inner.peers.insert(who.clone(), PeerData {
+			live: HashMap::new(),
+		});
+	}
+
+	fn peer_disconnected(&self, _context: &mut ValidatorContext<Block>, who: &PeerId) {
+		let mut inner = self.inner.write();
+		inner.peers.remove(who);
+	}
+
+	fn validate(&self, context: &mut ValidatorContext<Block>, sender: &PeerId, mut data: &[u8])
 		-> GossipValidationResult<Hash>
 	{
-		let orig_data = data;
-		match GossipMessage::decode(&mut data) {
-			Some(GossipMessage { relay_parent, statement }) => {
-				let live = self.live_session.read();
-				let topic = || ::router::attestation_topic(relay_parent.clone());
-				if let Some(validation) = live.get(&relay_parent) {
-					if validation.check_statement(&relay_parent, &statement) {
-						// repropagate
-						let topic = topic();
-						context.broadcast_message(topic, orig_data.to_owned(), false);
-						GossipValidationResult::ProcessAndKeep(topic)
-					} else {
-						GossipValidationResult::Discard
-					}
-				} else {
-					match self.oracle.is_known(&relay_parent) {
-						None | Some(Known::Leaf) => GossipValidationResult::ProcessAndKeep(topic()),
-						Some(Known::Old) | Some(Known::Bad) => GossipValidationResult::Discard,
+		let (res, cost_benefit) = match GossipMessage::decode(&mut data) {
+			None => (GossipValidationResult::Discard, cost::MALFORMED_MESSAGE),
+			Some(GossipMessage::Neighbor(VersionedNeighborPacket::V1(packet))) => {
+				let (res, cb, topics) = self.inner.write().validate_neighbor_packet(sender, packet);
+				for new_topic in topics {
+					context.send_topic(sender, new_topic, false);
+				}
+				(res, cb)
+			}
+			Some(GossipMessage::Statement(statement)) => {
+				let (res, cb) = self.inner.write().validate_statement(statement);
+				if let GossipValidationResult::ProcessAndKeep(ref topic) = res {
+					context.broadcast_message(topic.clone(), data.to_vec(), false);
+				}
+				(res, cb)
+			}
+		};
+
+		self.report(sender, cost_benefit);
+		res
+	}
+
+	fn message_expired<'a>(&'a self) -> Box<FnMut(Hash, &[u8]) -> bool + 'a> {
```

Comment:
the idea is that we return a closure with locks held so we don't lock on every message we check for expiry. we typically do all of them at once. otherwise, you need an associated type. The `GossipValidator` is stored as a trait object so we can't use an assoc. type